### PR TITLE
Update optaplanner version during release

### PR DESCRIPTION
**Pull request by dupliaka**
-  Updates optaplanner minor version in release branch
-  Force pushes to stable
-  Updates release micro version

Does not update development branch after release. 

Build reference: (will be available soon)

### JIRA

 https://issues.redhat.com/browse/PLANNER-2230

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: Jenkinsfile.deploy [hunk 1]
```diff
@@ -431,6 +456,6 @@ void setDeployPropertyIfNeeded(String key, def value) {
     }
 }
 
-String getQuickStartsBranch(String quickStartsBranch = 'development'){
-    return quickStartsBranch
+String getQuickStartsBranch(){
+    return isRelease()?getBuildBranch():params.QUICKSTARTS_BUILD_BRANCH_NAME
```
Comment (rsynek) [inline Jenkinsfile.deploy:460]: Formatting:
```suggestion
    return isRelease() ? getBuildBranch() : params.QUICKSTARTS_BUILD_BRANCH_NAME
```

File: Jenkinsfile.promote [hunk 2]
```diff
@@ -130,6 +131,23 @@ pipeline {
             }
         }
 
+        stage('Merge Quickstarts PR, tag and reset stage') {
```
Comment (rsynek) [inline Jenkinsfile.promote:134]:
```suggestion
        stage('Merge Quickstarts PR, tag and reset the stable branch') {
```

File: Jenkinsfile.promote [hunk 3]
```diff
@@ -386,7 +427,7 @@ void checkoutRepo(String repo, String branch) {
 }
 
 void checkoutRepo(String repo) {
-    checkoutRepo(repo, getBuildBranch())
+    checkoutRepo(repo, getBuildBranch() )
```
Comment (rsynek) [inline Jenkinsfile.promote:430]:
```suggestion
    checkoutRepo(repo, getBuildBranch())
```

File: Jenkinsfile.promote [hunk 4]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
     }
     mvnCmd.skipTests(true).run('clean deploy')
 }
+
+def updateAndResetBranch(String repo, String branchToReset){
```
Comment (rsynek) [inline Jenkinsfile.promote:518]:
```suggestion
def updateAndResetBranch(String repo, String branchToReset) {
```

File: Jenkinsfile.promote [hunk 5]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+        sh '''
+            sed -i \':a;N;$!ba;s/\\n *<repositories>.*<\\/repositories>//g\' */pom.xml
+            sed -i \':a;N;$!ba;s/\\n *<repositories>.*<\\/repositories>//g\' pom.xml
+            sed -i \':a;N;$!ba;s/\\n *\\/\\/ To get snapshots during development[^{}]*maven {[^{}]*mavenContent {[^{}]*snapshotsOnly[^{}]*}[^{}]*}//g\' */build.gradle
+        '''
+        assert !sh (script: 'grep -Rn "repository.jboss.org" --include={pom.xml,build.gradle}|cat'
+                , allowEmptyResults: true, returnStdout: true)
+}
+
+String getGhCredsID(){
```
Comment (rsynek) [inline Jenkinsfile.promote:540]:
```suggestion
String getGhCredsID() {
```

File: Jenkinsfile.promote [hunk 6]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+    return params.GITHUB_TOKEN_CREDS_ID;
+}
+
+String getProtectionMapScript(){
```
Comment (rsynek) [inline Jenkinsfile.promote:544]:
```suggestion
String getProtectionMapScript() {
```

File: Jenkinsfile.promote [hunk 7]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+        sh "rm -r protection.json"
+        if (enabled == allowForcePushEnabled) {
+            echo "[INFO] ${getGitAuthor()}/${repo}'s ${protectedBranch} force push enabled."
+        } else {
+            error "Couldn\'t change protection on ${protectedBranch}."
+        }
+    }
+}
+
+def forcePushProtectedBranch(String repo, String protectedBranch){
```
Comment (rsynek) [inline Jenkinsfile.promote:571]:
```suggestion
def forcePushProtectedBranch(String repo, String protectedBranch) {
```

File: Jenkinsfile.promote [hunk 8]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+    withCredentials([usernamePassword(credentialsId: getGitAuthorCredsID(), usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]) {
+        sh """
+        git config --local credential.helper \"!f() { echo username=\\$GIT_USERNAME; echo password=\\$GIT_PASSWORD; }; f\"
+        git push origin ${protectedBranch} --force
+        """
+    }
+    disableForcePushes(repo, protectedBranch)
+}
+
+void gradleVersionsUpdate(String newVersion){
```
Comment (rsynek) [inline Jenkinsfile.promote:582]:
```suggestion
void gradleVersionsUpdate(String newVersion) {
```

File: Jenkinsfile.promote [hunk 9]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+        assert !sh (script: 'grep -Rn "repository.jboss.org" --include={pom.xml,build.gradle}|cat'
+                , allowEmptyResults: true, returnStdout: true)
+}
+
+String getGhCredsID(){
+    return params.GITHUB_TOKEN_CREDS_ID;
+}
+
+String getProtectionMapScript(){
+    return "if .restrictions == null then . + {\"restrictions_used\":null}  else . + {\"restrictions_used\":{\"users\":[.restrictions.users[].login], \"team\":[.restrictions.users[].slug]}} end | {\"required_status_checks\": .required_status_checks,  \"required_pull_request_reviews\": {\"dismiss_stale_reviews\":(.required_pull_request_reviews.dismiss_stale_reviews//false), \"require_code_owner_reviews\":(.required_pull_request_reviews.require_code_owner_reviews//false)}, \"enforce_admins\":.enforce_admins.enabled, \"restrictions\":.restrictions_used}";
```
Comment (rsynek) [inline Jenkinsfile.promote:545]: Is it possible to make this line more readable (starting by breaking it into multiple lines)?

File: Jenkinsfile.promote [hunk 10]
```diff
@@ -470,3 +514,80 @@ void runMavenDeploy(MavenCommand mvnCmd) {
...
+def enableForcePushes(String repo, String protectedBranch) {
+    setAllowForcePushes(repo, protectedBranch, "true");
+}
+
+def disableForcePushes(String repo, String protectedBranch) {
+    setAllowForcePushes(repo, protectedBranch, "false");
+}
+
+def setAllowForcePushes(String repo, String protectedBranch, String enabled, String ghPath = "../gh") {
+    sh "if [ -x $ghPath ]; then echo 'gh found at $ghPath'; else echo 'gh not found at $ghPath' && exit 1; fi"
```
Comment (rsynek) [inline Jenkinsfile.promote:557]: Soft suggestion: create a new method called `assertGithubCLI`.

File: Jenkinsfile.promote (hunk 9 above)
Comment (rsynek) [reply Jenkinsfile.promote:545]: Let me just put down the concern we discussed: what if the JSON format changes? We don't want to silently lose any settings.

Comment (dupliaka) [reply Jenkinsfile.promote]: Added in 6ee6bd3 Thanks for the idea!

File: Jenkinsfile.promote (hunk 10 above)
Comment (dupliaka) [reply Jenkinsfile.promote]: Added 54ff8d1

**Review by rsynek — APPROVED** [review]
LGTM, nice work!

**Review by radtriste — APPROVED** [review]
Great job !

**Pull request merged**